Add Sonar Receiver

ellazzhang opened the pull request.

sampoornaravikanth committed:
Add json.org to the project

sampoornaravikanth committed:
Make SonarReciever

SonarReciever take sonar packets and stores them as simple ints.

ellazzhang committed:
Make sonarReciever only use one sonar sensor

porglezomp commented on src/org/usfirst/frc/team1076/robot/SonarReciever.java:
```diff
@@ -0,0 +1,37 @@
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+public class SonarReciever {
```
Please fix the spelling on this.

porglezomp commented on src/org/usfirst/frc/team1076/robot/SonarReciever.java:
```diff
@@ -0,0 +1,37 @@
+		byte[] buffer = new byte[512];
+		DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+		socket.receive(packet);
+		JSONTokener tokener = new JSONTokener(new String(packet.getData()));
```
You need to trim the `packet.getData()` to the size of the received data, this currently might include a bunch of junk at the end.

porglezomp commented on src/org/usfirst/frc/team1076/robot/SonarReciever.java:
```diff
@@ -0,0 +1,37 @@
+
+		byte[] buffer = new byte[512];
+		DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+		socket.receive(packet);
```
This is a blocking receive, if there's no data to receive this shouldn't block and should instead not change anything.

a2aaron committed:
Make SonarReceiver not block. Fix mispelled name. Add error handling for malformed JSON.

SonarReceiver no longer blocks when trying to receive data from a sensor.
If the data received is malformed or takes a long time to be sent, then
the value for the sensor is not updated.

Also Receiver is now spelled correctly.

porglezomp commented on src/org/usfirst/frc/team1076/robot/SonarReceiver.java:
```diff
@@ -0,0 +1,49 @@
+		DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+		socket.setSoTimeout(1);
+		try {
+		    socket.receive(packet);
```
What if this `receive` falls behind? We need to read all of the available packets.

a2aaron committed:
Make SonarReceiver not fall behind when receiving packets

If many packets pile up in the queue, then eventually
SonarReceiver will fall very far behind. Instead, we read 
from the packets until a timeout exception is thrown, 
indicating that we no longer have packets to read.

porglezomp reviewed (approved):
LGTM
